Disk Go check: Handle PartitionsWithContext errors gracefully instead of failing the check

jose-manuel-almaza opened the pull request:
###  What does this PR do?
Changes the disk check to handle errors from `PartitionsWithContext` gracefully, allowing the check to continue with whatever partitions were successfully retrieved instead of failing entirely.
### Motivation
`gopsutil.PartitionsWithContext` can return partial results along with an error when it fails to read some partitions but succeeds with others. Previously, the disk check would fail entirely on any error, causing a complete loss of disk metrics even when most partitions were accessible.
The Python implementation of the disk check handles this case correctly and continues collecting metrics for accessible partitions. Some customers had to roll back to the Python check as a workaround for this issue. This PR aligns the Go implementation with the Python behavior.
### Describe how you validated your changes
Updated existing test to reflect new behavior
Verified that no metrics are reported when no partitions are available
All existing tests pass
### Additional Notes
**Before:** Any error from `PartitionsWithContext` caused the check to return an error and report no metrics
**After:** Errors are logged as warnings, and the check continues with whatever partitions were retrieved. If no partitions are available, the check returns early without error.

jose-manuel-almaza committed:
fix(disk): handle PartitionsWithContext errors gracefully

- Don't fail the entire check if some partitions fail to load
- Log warning with number of partitions retrieved on error
- Early return if no partitions found

jose-manuel-almaza committed:
Add release note for disk check fix

nathan-b reviewed (commented):
I note that there's no actual test for the Windows case where partial results are returned to validate that the fix actually does what it's supposed to do. Ideally I'd like to see such a test to give us confidence rushing out this fix.

nathan-b reviewed (approved):
Given the urgency I'm going to go ahead and approve and trust you to evaluate the comments.

jose-manuel-almaza committed:
Improve tests: add test for partial partition results with error

jose-manuel-almaza committed:
Improve error handling: return error on complete failure, add test coverage

- Return error when PartitionsWithContext fails with no partitions (complete failure)
- Continue with partial results when some partitions are returned despite error
- Log warning when no partitions found without error (unusual case)
- Add test for 'no error + no partitions' scenario

jose-manuel-almaza commented on pkg/collector/corechecks/system/disk/diskv2/disk.go:
```diff
@@ -432,8 +432,11 @@ func (c *Check) collectPartitionMetrics(sender sender.Sender) error {
 	if err != nil {
-		log.Warnf("Unable to get disk partitions: %s", err)
-		return err
+		log.Warnf("Error getting some disk partitions (continuing with %d partitions): %v", len(partitions), err)
```
I was just working on covering all the different possible scenarios! I've also added tests for all of them.

jose-manuel-almaza committed:
Split partial results test into platform-specific files

- Move test from disk_test.go to disk_nix_test.go and disk_windows_test.go
- Use platform-appropriate device_name tag format in assertions
